library/std/src/sys/unix/fs/dir_fd.rs

```diff
@@ -0,0 +1,385 @@
+                    // just unlink files
+                    cvt(unsafe {
+                        unlinkat(current_readdir.as_fd().as_raw_fd(), child_name.as_ptr(), 0)
+                    })?;
+                } else {
+                    if let Some(child_readdir) =
+                        current_readdir.open_subdir_or_unlink_non_dir(child_name)?
+                    {
+                        // descend into this child directory
+
+                        let child_path_compoment =
+                            PathComponent::from_name_and_fd(child_name, child_readdir.as_fd())?;
+                        path_components.push(current_path_component);
+                        current_path_component = child_path_compoment;
+
+                        // avoid growing the cache over capacity
+                        if readdir_cache.len() == readdir_cache.capacity() {
+                            readdir_cache.pop_front();
+                        }
+                        readdir_cache.push_back(current_readdir);
+                        current_readdir = child_readdir;
+                    }
+                }
+            }
+
+            match path_components.pop() {
+                Some(parent_component) => {
+                    // going back up...
+
+                    // get parent directory readdir
+                    let parent_readdir = match readdir_cache.pop_back() {
+                        Some(readdir) => readdir,
+                        None => {
+                            // cache is empty - reopen parent
+                            let parent_readdir = current_readdir.get_parent()?;
+                            parent_component.verify_dev_ino(parent_readdir.as_fd())?;
+                            parent_readdir
+                        }
+                    };
+
+                    // remove now empty directory
+                    cvt(unsafe {
+                        unlinkat(
+                            parent_readdir.as_fd().as_raw_fd(),
+                            current_path_component.name.as_ptr(),
+                            libc::AT_REMOVEDIR,
+                        )
+                    })?;
+
+                    current_path_component = parent_component;
+                    current_readdir = parent_readdir;
+
+                    // Let "child directory" be the directory that was just deleted and "parent directory"
+                    // the parent of "child directory" which is now referred to in current_*.
+                    // If we don't have readdir open for the parent directory that means we got the file
+                    // descriptor via openat(dirfd, ".."). To make sure the that the child directory
+                    // was not moved somewhere else and the parent just happens to have the same reused
+                    // (dev, inode) pair, that we found descending, we check the parent directory
+                    // (dev, inode) as well.
```

Comment:
Shouldn't this check happen before we do the unlinking?

Reply:
Correct, fixed.

Reply:
Yes, fixed.